realtek: fix link status detection on RTL83xx

bkobl opened the pull request:
In combination with some PHYs, in particular the RTL8214FC, the
link state needs to be read 2x, because of a latching or latency
effect.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Hurricos commented:
Hey @bkobl, is there any chance you have access to Realtek documentation about that latching effect on e.g. RTL8214FC?

bkobl commented:
> Hey @bkobl, is there any chance you have access to Realtek documentation about that latching effect on e.g. RTL8214FC?

No, I don't. I also don't think this is due to the PHY itself, but because of how the PHYs are being polled by the SoC, which again depends on the PHY. The link state information is collected by the SoC (the ASIC, not the CPU) by polling the PHY via MDIO. While interrupts are being reliably fired when the link state changes, the SoC's link link state register at this point is not necessarily correct. So the PHY gave the right information, but the SoC does not always automatically (or in time) update the SoC link state register. There are cases where the information is correct, others where phylink compensates by reading the PHY itself and others where there is an issue. But the right solution is to always read the link-state register 2x.

Hurricos commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl839x.c:
```diff
@@ -611,6 +611,8 @@ irqreturn_t rtl839x_switch_irq(int irq, void *dev_id)
+	link = rtl839x_get_port_reg_le(RTL839X_MAC_LINK_STS);
 	for (i = 0; i < RTL839X_CPU_PORT; i++) {
 		if (ports & BIT_ULL(i)) {
 			link = rtl839x_get_port_reg_le(RTL839X_MAC_LINK_STS);
```
Looks to me that for rtl839x, we are now probing (`rtl839x_get_port_reg_le(RTL839X_MAC_LINK_STS)`'ing) *thrice*, not twice.

bkobl replied:
Oops, I tried to be too thorough. I'll fix that. Thanks for spotting this!

Hurricos commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl839x.c:
```diff
@@ -611,6 +611,8 @@ irqreturn_t rtl839x_switch_irq(int irq, void *dev_id)
 	pr_debug("RTL8390 Link change: status: %x, ports %llx\n", status, ports);
 
+	link = rtl839x_get_port_reg_le(RTL839X_MAC_LINK_STS);
+	link = rtl839x_get_port_reg_le(RTL839X_MAC_LINK_STS);
```
Edit: Or, well, at least thrice. I think only one of these would be needed.

Commit:
realtek: fix link status detection on RTL838x

In combination with some PHYs, in particular the RTL8214FC, the
link state needs to be read 2x, because of a latching or latency
effect.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

bkobl commented:
I fixed the issue above by not adding further LINK status reads on RTL839x.

Neustradamus commented:
@Hurricos: Now the PR author account is a ghost, what is the situation now for this PR?
